### ISSUE TITLE
SSR support with no memory leaks

### ISSUE THREAD

**Issue by elhoyos** [splitio/react-client#11]
I've been following @NicoZelaya's `useClient` suggestion to solve the memory leak reported in https://github.com/splitio/react-client/pull/2 without success.

I have a next.js@9.3.1 application and I want to take advantage of SSR without using Redux, but I'm getting a console message at server-time, with the additional downside that no children component is rendered in the client-side:

```sh
Shared Client not supported by the storage mechanism. Create isolated instances instead.
```

This HOC wraps `App`:

```tsx
// in withSplit.tsx
const unmountSplitClient = (client: SplitIO.IClient | null) => {
  if (client) {
    client.destroy();
  }
}

// Need a class to implement componentWillUnmount
class ServerSplitDestroyer extends React.Component {
  private client: SplitIO.IClient;

  constructor(props: { client: SplitIO.IClient }) {
    super(props);
    this.client = props.client;
  }

  // server-side unmount
  componentWillUnmount() {
    unmountSplitClient(this.client);
  }

  render() {
    return (<></>)
  }
}

const ClientDestroyer: NextPage<AType> = ({ splitioKey, children }) => {
  const client = useClient(splitioKey);

  // client-side unmount
  React.useEffect(() => {
    return function _unmountSplitClient() {
      unmountSplitClient(client);
    }
  }, [client]);

  return (
    <>
    <ServerSplitDestroyer />
    {children}
    </>
  );
}

export function withSplit() {
  return function withSplitFactoryHOC<OuterProps>(WrappedComponent: React.ComponentType<OuterProps>) {
    return (props: OuterProps) => {
      return (
        <SplitFactory config={setConfig(key)} >
          {(splitProps) => {
            return (
              <ClientDestroyer splitioKey={key}>
                <WrappedComponent {...props} {...splitProps} />
              </ClientDestroyer>
            )
          }}
        </SplitFactory>
      )
    }
  }
}
```

How can I use split.io in a next.js application with SSR support?

**Comment by elhoyos** [comment]
I should highlight that I do not experience any memory leak in the browser after commenting the `ServerSplitDestroyer` usage.

**Comment by elhoyos** [comment]
Interestingly, I'm unable to reproduce the client-side leak anymore. I could reproduce it yesterday without the `client.destroy`.

The server-side leak remains, though.

**Comment by chillaq** [comment]
Hi, the Split react SDK can only be loaded in the client side, not sure why you are seeing server-side leaks, are you loading the SDK on the server side as well?

Thanks
Bilal

**Comment by chillaq** [comment]
Hi @elhoyos , just want to follow up on this issue, did you have the chance to review the last comment? do you still need any help on this issue?

Thanks
Bilal

**Comment by NicoZelaya** [comment]
Hi @elhoyos ! pinging you before closing for inactivity. Let us know if you still need any assistance, happy to have a discussion on how you can continue if still having issues.

Best,
Nico.

**Comment by elhoyos** [comment]
Thank you all for the heads-up. Let me try to rep. and isolate the issue this weekend.

**Comment by elhoyos** [comment]
@chillaq & @NicoZelaya, here's an isolated playground that shows how to make it leak: https://github.com/elhoyos/next-split-ssr-leak#leak-1.

In short, you need to visit various pages to see "split" objects accumulating.

Baseline dump:
![baseline dump](https://user-images.githubusercontent.com/577520/88576486-8276c800-d013-11ea-9b15-904204b46252.png)

Last dump:
![last dump](https://user-images.githubusercontent.com/577520/88576521-8d315d00-d013-11ea-9c51-3b4db7c67024.png)

**Comment by elhoyos** [comment]
To shed a bit more of light:
* After some time, the development server flushes the cache and a heapdump will not show any split object
* I only see a couple of `SplitError` object appearing when I use a production-like bundle created with `npm run build`. Yet, I'm not sure whether this is because there is no leak at all or because the build obfuscates the other "split" objects.

**Comment by elhoyos** [comment]
On the other hand, you can see the message `Shared Client not supported by the storage mechanism. Create isolated instances instead.` when launching [tag/shared-client-not-supported](https://github.com/elhoyos/next-split-ssr-leak/releases/tag/shared-client-not-supported), while attempting to `client.destroy()`.

**Comment by NicoZelaya** [comment]
Thank you! I'll try to take a look. Regardless, as Bilal stated before, the React SDK is for the client side (browser) but not for running on a Node platform (like when you perform server side rendering).

If you want to use the SDK in both sides, you need to use the regular Split SDK and control the API for [Node](https://help.split.io/hc/en-us/articles/360020564931-Node-js-SDK) when processing a template on the server and only if you need to use feature flagging there, then do the same on the JavaScript you execute on the Browser with the [client side API](https://help.split.io/hc/en-us/articles/360020448791-JavaScript-SDK).

We have other plans for the future but the log you get is correct and is telling you that what you're doing, is not supported.
> Shared Client not supported by the storage mechanism. Create isolated instances instead

I'll take a look at your code later today or tomorrow and comment regardless, but it sounds that it'll be best for you to not use the React SDK and actually the plain JavaScript SDK in either or both of it's variants. 

If you need help figuring out how to do that, let's communicate through support@split.io so we can schedule a zoom meeting and discuss how to best integrate in your use case.

Best,
Nico.

**Comment by masad-frost** [comment]
What was the resolution for this ticket? We're migrating to split and this is a blocker for us. Did you end up passing the node factory somehow during SSR?

**Comment by EmilianoSanchez** [comment]
Hi @elhoyos , @masad-frost ,

I wanted to share with you, that we have released React SDK v1.11.0, which includes a new `SplitFactoryProvider` component as a drop-in replacement of the `SplitFactory` component, now deprecated.

The new component fixes the memory leaks and issues of the `SplitFactory` component when using it on server side or with React development and strict modes.

`SplitFactory` will be removed eventually in the next major release (v2.0.0).
You can check this [issue #109 comment](https://github.com/splitio/react-client/issues/109#issuecomment-1894362198) for details and some examples to use it for server-side rendering and NextJS.

I will close the issue for now but don't hesitate to reopen it or create a new one if you encounter any further problems or additional questions.
